### PR TITLE
Support for Custom Validator Message

### PR DESCRIPTION
**Pull request by brianwyka**
This allows for custom validator implementations to override the message template of a given annotation and to supply its own to the context.

This can be useful if validation is at class-level and is validating multiple fields, where the validation message needs to be different.

### CODE REVIEW

File: inject-java/src/main/java/io/micronaut/annotation/processing/AnnotationProcessingOutputVisitor.java [hunk 1]
```diff
@@ -47,24 +47,24 @@
...
+            try {
+                addOpenJavaModules(FilterOutputStream.class, AnnotationProcessingOutputVisitor.class);
+                field.setAccessible(true);
+                return field;
+            } catch (Exception e) {
+                return null;
+            }
+        })
+        .orElse(null);
+
```
Comment (brianwyka) [inline inject-java/src/main/java/io/micronaut/annotation/processing/AnnotationProcessingOutputVisitor.java:61]: Unrelated checkstyle fix...

File: runtime/src/main/java/io/micronaut/jackson/modules/BeanIntrospectionModule.java [hunk 2]
```diff
@@ -55,7 +73,14 @@
...
 import java.lang.annotation.Annotation;
-import java.util.*;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
```
Comment (brianwyka) [inline runtime/src/main/java/io/micronaut/jackson/modules/BeanIntrospectionModule.java:83]: Unrelated checkstyle fix...

File: validation/src/main/java/io/micronaut/validation/validator/constraints/ConstraintValidatorContext.java [hunk 3]
```diff
@@ -48,5 +48,21 @@
      * @return The root bean under validation.
      */
     @Nullable Object getRootBean();
+
+    /**
+     * Sets a message template to be used for the validation error to override default.
+     *
+     * @param messageTemplateOverride the message to override default
+     */
+    default void overrideMessageTemplate(@Nullable final String messageTemplateOverride) { }
```
Comment (jameskleeh) [inline validation/src/main/java/io/micronaut/validation/validator/constraints/ConstraintValidatorContext.java]: this should throw UnsupportedOperationException

**Review by jameskleeh — CHANGES_REQUESTED** [review]
Given this is a new feature it should target the 2.5.x branch. I'll let others weigh in on whether this approach is a good idea or not

**Comment by brianwyka** [comment]
> Given this is a new feature it should target the 2.5.x branch. I'll let others weigh in on whether this approach is a good idea or not

Changed base branch to `2.5.x`

File: validation/src/main/java/io/micronaut/validation/validator/constraints/ConstraintValidatorContext.java (hunk 3 above)
Comment (brianwyka) [reply validation/src/main/java/io/micronaut/validation/validator/constraints/ConstraintValidatorContext.java]: Done.

**Comment by brianwyka** [comment]
@jameskleeh, who are the others here? Should we tag them for review?

File: validation/src/main/java/io/micronaut/validation/validator/constraints/ConstraintValidatorContext.java [hunk 4]
```diff
@@ -48,5 +48,23 @@
      * @return The root bean under validation.
      */
     @Nullable Object getRootBean();
+
+    /**
+     * Sets a message template to be used for the validation error to override default.
+     *
+     * @param messageTemplateOverride the message to override default
+     */
+    default void overrideMessageTemplate(@Nullable final String messageTemplateOverride) {
```
Comment (graemerocher) [inline validation/src/main/java/io/micronaut/validation/validator/constraints/ConstraintValidatorContext.java]: Maybe just call this `messageTemplate`, explicitly prefixing with `override` seems an implementation detail

File: validation/src/main/java/io/micronaut/validation/validator/constraints/ConstraintValidatorContext.java [hunk 5]
```diff
@@ -48,5 +48,23 @@
      * @return The root bean under validation.
      */
     @Nullable Object getRootBean();
+
+    /**
+     * Sets a message template to be used for the validation error to override default.
+     *
+     * @param messageTemplateOverride the message to override default
```
Comment (graemerocher) [inline validation/src/main/java/io/micronaut/validation/validator/constraints/ConstraintValidatorContext.java]: This method needs `@since 2.5.0`

File: validation/src/main/java/io/micronaut/validation/validator/constraints/ConstraintValidatorContext.java [hunk 6]
```diff
@@ -48,5 +48,23 @@
...
+    default void overrideMessageTemplate(@Nullable final String messageTemplateOverride) {
+        throw new UnsupportedOperationException("not implemented");
+    }
+
+    /**
+     * Get the message template override.
+     *
+     * @return the message override
+     */
+    @Nullable default String getMessageTemplateOverride() {
```
Comment (graemerocher) [inline validation/src/main/java/io/micronaut/validation/validator/constraints/ConstraintValidatorContext.java]: Does this need to be `public` API? Seems it is only used internally. Either way it should return `Optional` and if it must be public be annotated with `@since 2.5.0`

**Review by graemerocher — CHANGES_REQUESTED** [review]

**Comment by brianwyka** [comment]
Ready for re-review @graemerocher

**Review by graemerocher — APPROVED** [review]
